sql/auth/sql_user_table.cc

```diff
@@ -2026,15 +2038,39 @@ int open_grant_tables(THD *thd, TABLE_LIST *tables,
     Perform the TOI after the replication filter check to avoid
     replicating commands that won't be applied locally (due to a filter).
   */
-  /* Doing TOI here is not the ideal solution, as we are holding
-  MDL locks already and the current thread can still be BF-aborted.
-  But we have to check replication filters before TOI and for this
-  we need MDL locks. */
   if (WSREP(thd) && !skip_toi &&
       wsrep_to_isolation_begin(thd, db, table, nullptr)) {
     WSREP_ERROR("Fail to replicate: %s", thd->query().str);
     return -1;
   }
+
+  /* For replica thread we checked replication filters before TOI and then
+     released MDL locks.
+     Original flow continues with tables locked, so lock them again. */
+  if (acl_tables_setup_for_write_and_acquire_mdl(thd, tables)) return -1;
```

Comment:
Can we have this logic of temporary release of lock and re-acquire to apply only for WSREP enabled threads?
```suggestion
  if (WSREP(thd) && acl_tables_setup_for_write_and_acquire_mdl(thd, tables)) return -1;
```

Reply:
Please imagine the flow for non-slave thread. 
We have to lock always.